tests/utils/test_operator.py

```diff
@@ -13,15 +13,178 @@
 
 import pytest
 from ruamel.yaml import YAML
-from ruamel.yaml.comments import CommentedMap
+from ruamel.yaml.comments import CommentedMap, CommentedSeq
 
 from atomic_reactor.util import ImageName, chain_get
-from atomic_reactor.utils.operator import OperatorCSV, OperatorManifest, NotOperatorCSV
+from atomic_reactor.utils.operator import (
+    OperatorCSV,
+    OperatorManifest,
+    NotOperatorCSV,
+    default_pullspec_heuristic,
+)
 
 
 yaml = YAML()
 
 
+SHA = "5d141ae1081640587636880dbe8489439353df883379158fa8742d5a3be75475"
+
+
+@pytest.mark.parametrize("text, expected", [
+    # Trivial cases
+    ("a.b/c:1", ["a.b/c:1"]),
+    ("a.b/c/d:1", ["a.b/c/d:1"]),
+    # Digests in tag
+    ("a.b/c@sha256:{sha}".format(sha=SHA), ["a.b/c@sha256:{sha}".format(sha=SHA)]),
+    ("a.b/c/d@sha256:{sha}".format(sha=SHA), ["a.b/c/d@sha256:{sha}".format(sha=SHA)]),
+    # Port in registry
+    ("a.b:1/c:1", ["a.b:1/c:1"]),
+    ("a.b:5000/c/d:1", ["a.b:5000/c/d:1"]),
+    # Special characters everywhere
+    ("a-b.c_d/e-f.g_h/i-j.k_l@sha256:{sha}".format(sha=SHA),
+     ["a-b.c_d/e-f.g_h/i-j.k_l@sha256:{sha}".format(sha=SHA)]),
+    ("a-._b/c-._d/e-._f:g-._h", ["a-._b/c-._d/e-._f:g-._h"]),
+    # Multiple namespaces
+    ("a.b/c/d/e:1", ["a.b/c/d/e:1"]),
+    ("a.b/c/d/e/f/g/h/i:1", ["a.b/c/d/e/f/g/h/i:1"]),
+    # Enclosed in various non-pullspec characters
+    (" a.b/c:1 ", ["a.b/c:1"]),
+    ("\na.b/c:1\n", ["a.b/c:1"]),
+    ("\ta.b/c:1\t", ["a.b/c:1"]),
+    (",a.b/c:1,", ["a.b/c:1"]),
+    (";a.b/c:1;", ["a.b/c:1"]),
+    ("'a.b/c:1'", ["a.b/c:1"]),
+    ('"a.b/c:1"', ["a.b/c:1"]),
+    ("<a.b/c:1>", ["a.b/c:1"]),
+    ("`a.b/c:1`", ["a.b/c:1"]),
+    ("*a.b/c:1*", ["a.b/c:1"]),
+    ("(a.b/c:1)", ["a.b/c:1"]),
+    ("[a.b/c:1]", ["a.b/c:1"]),
+    ("{a.b/c:1}", ["a.b/c:1"]),
+    # Enclosed in various pullspec characters
+    (".a.b/c:1.", ["a.b/c:1"]),
+    ("-a.b/c:1-", ["a.b/c:1"]),
+    ("_a.b/c:1_", ["a.b/c:1"]),
+    ("/a.b/c:1/", ["a.b/c:1"]),
+    ("@a.b/c:1@", ["a.b/c:1"]),
+    (":a.b/c:1:", ["a.b/c:1"]),
```

Comment:
`8.0-1136.1589201512` can we also test tag like this?

Reply:
Sure,  the "special characters everywhere" cases are kind of similar but I'll add a case